From 4933da8d3c149ba163fd87db44f12197da672b58 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Mon, 14 Jul 2014 13:08:04 -0400
Subject: [idle] intel-idle: Fix CPU hotplug callback registration

Message-id: <1405343292-21004-28-git-send-email-prarit@redhat.com>
Patchwork-id: 85771
O-Subject: [RHEL7.1 PATCH BZ 1119078 27/35] intel-idle: Fix CPU hotplug callback registration
Bugzilla: 1119078
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Igor Mammedov <imammedo@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1119078

commit 07494d547e92bde6857522d2a92ff70896aecadb
Author: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
Date:   Tue Mar 11 02:10:30 2014 +0530

    intel-idle: Fix CPU hotplug callback registration

    Subsystems that want to register CPU hotplug callbacks, as well as perform
    initialization for the CPUs that are already online, often do it as shown
    below:

     get_online_cpus();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     register_cpu_notifier(&foobar_cpu_notifier);

     put_online_cpus();

    This is wrong, since it is prone to ABBA deadlocks involving the
    cpu_add_remove_lock and the cpu_hotplug.lock (when running concurrently
    with CPU hotplug operations).

    Instead, the correct and race-free way of performing the callback
    registration is:

     cpu_notifier_register_begin();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     /* Note the use of the double underscored version of the API */
     __register_cpu_notifier(&foobar_cpu_notifier);

     cpu_notifier_register_done();

    Fix the intel-idle code by using this latter form of callback registration.

    Cc: Len Brown <lenb@kernel.org>
    Cc: Ingo Molnar <mingo@kernel.org>
    Signed-off-by: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Don Zickus <dzickus@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/idle/intel_idle.c b/drivers/idle/intel_idle.c
index d47bb0f..1fcf905 100644
--- a/drivers/idle/intel_idle.c
+++ b/drivers/idle/intel_idle.c
@@ -692,14 +692,19 @@ static int __init intel_idle_init(void)
  if (intel_idle_cpuidle_devices == NULL)
   return -ENOMEM;
 
+ cpu_notifier_register_begin();
+
  for_each_online_cpu(i) {
   retval = intel_idle_cpu_init(i);
   if (retval) {
+   cpu_notifier_register_done();
    cpuidle_unregister_driver(&intel_idle_driver);
    return retval;
   }
  }
- register_cpu_notifier(&cpu_hotplug_notifier);
+ __register_cpu_notifier(&cpu_hotplug_notifier);
+
+ cpu_notifier_register_done();
 
  return 0;
 }
@@ -709,10 +714,13 @@ static void __exit intel_idle_exit(void)
  intel_idle_cpuidle_devices_uninit();
  cpuidle_unregister_driver(&intel_idle_driver);
 
+ cpu_notifier_register_begin();
 
  if (lapic_timer_reliable_states != LAPIC_TIMER_ALWAYS_RELIABLE)
   on_each_cpu(__setup_broadcast_timer, (void *)false, 1);
- unregister_cpu_notifier(&cpu_hotplug_notifier);
+ __unregister_cpu_notifier(&cpu_hotplug_notifier);
+
+ cpu_notifier_register_done();
 
  return;
 }
-- 
1.7.1